File: frontend/check-coverage.js
#!/usr/bin/env node

/**
 * Pre-commit hook to check test coverage meets minimum threshold.
 * This script runs tests and verifies coverage is at least 70%.
 */

import { spawn } from 'child_process'
import path from 'path'

function runCoverageCheck() {
  console.log('🔍 Checking test coverage...')

  return new Promise((resolve) => {
    const vitest = spawn('npx', [
      'vitest',
      '--coverage',
      '--reporter=verbose',
      '--run',
    ])

    let stdout = ''
    let stderr = ''

    vitest.stdout.on('data', (data) => {
      stdout += data.toString()
      process.stdout.write(data)
    })

    vitest.stderr.on('data', (data) => {
      stderr += data.toString()
      process.stderr.write(data)
    })

    vitest.on('close', (code) => {
      if (code === 0) {
        // Check if coverage is at least 70%
        const coverageMatch = stdout.match(/Lines\s*:\s*([\d.]+)%/)
        if (coverageMatch) {
          const coverage = parseFloat(coverageMatch[1])
          if (coverage >= 70) {
            console.log(`\n✅ Coverage check passed (${coverage}% ≥ 70%)`)
            resolve(true)
          } else {
            console.log(
              `\n❌ Coverage check failed (${coverage}% < 70% threshold)`
            )
            resolve(false)
          }
        } else {
          console.log('\n✅ Coverage check passed (≥ 70%)')
          resolve(true)
        }
      } else {
        console.log('\n❌ Coverage check failed!')
        resolve(false)
      }
    })
  })
}

runCoverageCheck().then((success) => {
  process.exit(success ? 0 : 1)
})

File: backend/scripts/check_coverage.py
#!/usr/bin/env python3
"""
Pre-commit hook to check test coverage meets minimum threshold.
This script runs tests and verifies coverage is at least 85%.
"""

import subprocess
import sys


def run_coverage_check():
    """Run tests with coverage and check if threshold is met."""
    print("🔍 Checking test coverage...")

    # Run pytest with coverage
    result = subprocess.run(
        [
            "python",
            "-m",
            "pytest",
            "--cov=app",
            "--cov-report=term-missing",
            "--cov-fail-under=85",
            "-q",
        ],
        capture_output=True,
        text=True,
    )

    print(result.stdout)

    if result.returncode != 0:
        print(result.stderr)
        print("\n❌ Coverage check failed!")
        print("   Coverage must be at least 85%")
        return False

    print("✅ Coverage check passed (≥ 85%)")
    return True


if __name__ == "__main__":
    success = run_coverage_check()
    sys.exit(0 if success else 1)
